### PR TITLE
Correct warm-up with fp8

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

This PR starts the work of having fine-grained allocation when using quantizers. For now, it reflects the actual library behavior, while simply adjusting fp8 quantization (for Deepseek), but it should become more general in the future (i.e. probably the default should be 2 if most quantizers correctly pre-process? Did not check all for now), and be overriden only for a few.

cc @SunMarc @MekkCyber

### CODE REVIEW

File: src/transformers/quantizers/base.py [hunk 1]
```diff
@@ -252,6 +252,17 @@ def dequantize(self, model):
 
         return model
 
+    def get_cuda_warm_up_factor(self):
+        """
+        The factor to be used in `caching_allocator_warmup` to get the number of bytes to pre-allocate to warm up cuda.
+        A factor of 2 means we allocate all bytes in the empty model (since we allocate in fp16), a factor of 4 means
```
Comment (SunMarc) [inline src/transformers/quantizers/base.py:258]: Doesn't this depend on the value that we passed in `from_pretrained` -> `torch_dtype` since you have the following ?  
`param_byte_count = param.numel() * param.element_size()`

For example, if you load the model in fp16, with a factor of 2, we will only allocate half of the required bytes (since you set dtype = torch.float16) in the empty model ? 

`_ = torch.empty(byte_count // factor, dtype=torch.float16, device=device, requires_grad=False)`

**Review by SunMarc — APPROVED** [review]
Thanks for the fix ! Just a small question to better understand

File: src/transformers/quantizers/base.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/quantizers/base.py:258]: Nop we still allocate all bytes that the model will use. Basically, at this point the model has meta parameters, but with the correct dtype, so `param_byte_count = param.numel() * param.element_size()` adds up correctly all the bytes that will be needed. And then, 
, `_ = torch.empty(byte_count // factor, dtype=torch.float16, device=device, requires_grad=False)` allocates all those bytes (we just divide by 2 because fp16 will take-up 2 bytes per param, so the total allocated is still `byte_count`)

Comment (Cyrilvallez) [reply src/transformers/quantizers/base.py:258]: But we could divide by 4 and allocate with fp32, this one is arbitrary -> we just need to allocate all `byte_count`

Comment (MekkCyber) [reply src/transformers/quantizers/base.py:258]: why not using torch.int8 then ? is there a specific reason ?

Comment (SunMarc) [reply src/transformers/quantizers/base.py:258]: Oh makes sense, thanks for explaining !

Comment (Cyrilvallez) [reply src/transformers/quantizers/base.py:258]: I just was not sure `empty` would fit well with int dtypes (i.e. we really don't want torch to have to perform casting operations except than simply allocating the memory)- so I chose fp16, but otherwise no reason
